test/shared/src/main/scala/zio/test/Spec.scala

```diff
@@ -16,183 +16,277 @@
+      case TestCase(label, test)         => TestCase(label, test)
+    }
+
   /**
    * Computes the size of the spec, i.e. the number of tests in the spec.
    */
-  final def size: Int = fold[Int] {
-    case SuiteCase(_, counts, _) => counts.sum
-    case TestCase(_, _)          => 1
+  final def size: ZIO[R, E, Int] = fold[ZIO[R, E, Int]] {
+    case SuiteCase(_, counts, _) => counts.flatMap(ZIO.collectAll(_).map(_.sum))
+    case TestCase(_, _)          => ZIO.succeed(1)
   }
 
   /**
    * Transforms the spec one layer at a time.
    */
-  final def transform[L1, T1](f: SpecCase[L, T, Spec[L1, T1]] => SpecCase[L1, T1, Spec[L1, T1]]): Spec[L1, T1] =
+  final def transform[R1, E1, L1, T1](
+    f: SpecCase[R, E, L, T, Spec[R1, E1, L1, T1]] => SpecCase[R1, E1, L1, T1, Spec[R1, E1, L1, T1]]
+  ): Spec[R1, E1, L1, T1] =
     caseValue match {
-      case SuiteCase(label, specs, exec) =>
-        Spec(f(SuiteCase(label, specs.map(_.transform(f)), exec)))
-      case t @ TestCase(_, _) => Spec(f(t))
+      case SuiteCase(label, specs, exec) => Spec(f(SuiteCase(label, specs.map(_.map(_.transform(f))), exec)))
+      case t @ TestCase(_, _)            => Spec(f(t))
     }
 
   /**
    * Transforms the spec statefully, one layer at a time.
    */
-  final def transformAccum[L1, T1, Z](
+  final def transformAccum[R1, E1, L1, T1, Z](
     z0: Z
-  )(f: (Z, SpecCase[L, T, Spec[L1, T1]]) => (Z, SpecCase[L1, T1, Spec[L1, T1]])): (Z, Spec[L1, T1]) =
+  )(
+    f: (Z, SpecCase[R1, E, L, T, Spec[R1, E1, L1, T1]]) => (Z, SpecCase[R1, E1, L1, T1, Spec[R1, E1, L1, T1]])
+  ): ZIO[R, E, (Z, Spec[R1, E1, L1, T1])] =
     caseValue match {
       case SuiteCase(label, specs, exec) =>
-        val (z, specs1) =
-          specs.foldLeft(z0 -> Vector.empty[Spec[L1, T1]]) {
-            case ((z, vector), spec) =>
-              val (z1, spec1) = spec.transformAccum(z)(f)
-
-              z1 -> (vector :+ spec1)
-          }
-
-        val (z1, caseValue) = f(z, SuiteCase(label, specs1, exec))
-
-        z1 -> Spec(caseValue)
+        for {
+          specs <- specs
+          result <- ZIO.foldLeft(specs)(z0 -> Vector.empty[Spec[R1, E1, L1, T1]]) {
+                     case ((z, vector), spec) =>
+                       spec.transformAccum(z)(f).map { case (z1, spec1) => z1 -> (vector :+ spec1) }
+                   }
+          (z, specs1)     = result
+          res             = f(z, SuiteCase(label, ZIO.succeed(specs1), exec))
+          (z1, caseValue) = res
+        } yield z1 -> Spec(caseValue)
       case t @ TestCase(_, _) =>
         val (z, caseValue) = f(z0, t)
-        z -> Spec(caseValue)
+        ZIO.succeed(z -> Spec(caseValue))
     }
 }
 object Spec {
-  sealed trait SpecCase[+L, +T, +A] { self =>
-    final def map[B](f: A => B): SpecCase[L, T, B] = self match {
-      case SuiteCase(label, specs, exec) => SuiteCase(label, specs.map(f), exec)
+  sealed trait SpecCase[-R, +E, +L, +T, +A] { self =>
```

Comment:
Ah, I see you propagated `R` and `E` into `SpecCase`. A lot of type parameters but inference should be extremely good due to lower order nature + variance annotations.

What the design really wants to be is:

```scala
sealed trait SpecCase[F[_], +L, +T, +A]
```

So we can have:

```scala
final case class Effect[F[_], +L, +T, +A](fa: F[A]) extends SpecCase[F, L, T, A]
```

where in a common case, `F[A] = ZIO[R, E, A]`. And in another case, `F[A] = A`.

The problem is ZIO eschews higher-kinded types for teachability / inference, so doing so here would be rather inconsistent with the rest of ZIO, and is almost certain to diminish inference. Although I suppose we could test it.

Reply:
Yes I was thinking the same thing but if we managed to do functional effects and streaming without higher order types it seems like we should be able to do make it work for testing. 😃

Reply:
True, we should keep trying. 😄